File: client-sdk/typescript/src/context.ts
// `nats` CLI context loader and URL parser.
//
// Two entry points produce `@nats-io/transport-node` connection options:
//
//   - {@link loadContextOptions} — reads context files written by
//     `nats context add` / `nats context select` under `~/.config/nats`
//     (or $NATS_CONFIG_HOME / $XDG_CONFIG_HOME/nats).
//   - {@link parseNatsUrl} — parses a single NATS URL and extracts
//     credentials from `userinfo` if present (token, or user:pass).
//     The bare `@nats-io/transport-node` `connect({ servers })` does NOT
//     parse userinfo — it expects credentials as separate config fields —
//     but the `nats` CLI does, which causes a confusing UX gap. Use this
//     helper to bridge the two.
//
// Both return `NodeConnectionOptions` you can pass straight to `connect()`:
//
//     import { connect } from "@nats-io/transport-node";
//     import { Agents, loadContextOptions, parseNatsUrl } from "@synadia-ai/agents";
//
//     const opts = await loadContextOptions("prod");
//     // or:
//     const opts = parseNatsUrl("nats://TOKEN@nats.example.com:4222");
//
//     const nc = await connect(opts);
//     const agents = new Agents({ nc });
//
// Supported context fields: `url`, `creds` (path), `nkey` (path),
// `user_jwt` (+ optional `user_seed` for nonce signing), `user`+`password`,
// `token`, `inbox_prefix`, plus the TLS file triple `cert`/`key`/`ca` and
// `tls_first`.
// Skipped: `nsc` integration.
//
// File-path fields (`creds`, `nkey`, and the TLS triple) expand a leading
// `~/` to the home directory; any other relative path resolves against the
// process working directory (as the `nats` CLI does), not the context
// file's directory.
//
// Precedence inside a context: `creds` > `nkey` > `user_jwt` (+`user_seed`
// when present) > `user`/`password` > `token`.
//
// `user_jwt` without an accompanying nkey seed leaves the CONNECT signature
// empty, so it only works against servers that do not require nonce signing.

import { readFile } from "node:fs/promises";
import { homedir } from "node:os";
import { join } from "node:path";
import { credsAuthenticator, jwtAuthenticator, nkeyAuthenticator } from "@nats-io/nats-core";
import type { NodeConnectionOptions } from "@nats-io/transport-node";
import { NatsContextError } from "./errors.js";

/** A NATS CLI context file, read but not yet turned into connection options. */
export interface NatsContextFile {
  /** The resolved context name (after `"current"` resolution). */
  readonly name: string;
  /** Absolute path of the context JSON file. */
  readonly path: string;
  /** The parsed JSON object, verbatim. */
  readonly fields: Readonly<Record<string, unknown>>;
}

/**
 * Auth material selected from one context-file read. Internal contract for
 * `resolveNatsConnectionBundle`; raw bytes are never exported from the
 * package root or included in logs/errors.
 *
 * @internal
 */
export type ContextAuthSnapshot =
  | { readonly kind: "creds"; readonly bytes: Uint8Array }
  | { readonly kind: "nkey"; readonly bytes: Uint8Array }
  | { readonly kind: "jwt-seed"; readonly jwt: string; readonly seed: Uint8Array }
  | { readonly kind: "jwt" }
  | { readonly kind: "user-password" }
  | { readonly kind: "token" }
  | { readonly kind: "none" };

/**
 * Connection options plus the exact auth snapshot used to build their
 * authenticator. Internal so the connection-bundle helper can derive a
 * signer without re-reading a context or credential file.
 *
 * @internal
 */
export interface ContextConnectionSnapshot {
  readonly name: string;
  readonly connectionOptions: NodeConnectionOptions;
  readonly auth: ContextAuthSnapshot;
  /** Zero retained auth bytes. Call only after the connection is closed. */
  wipe(): void;
}

/**
 * Read a NATS CLI context by name. Pass `"current"` to resolve via
 * `$NATS_CONTEXT` or the `context.txt` selection file. This is the
 * *reading* half of {@link loadContextOptions}; `signerFromContext` in the
 * identity module reuses it to find the seed / creds path without
 * building connection options.
 */
export async function readContextFile(selector: string): Promise<NatsContextFile> {
  const baseDir = resolveBaseDir();
  const name = selector === "current" ? await resolveCurrentName(baseDir) : selector;

  if (name.includes("/") || name.includes("\\") || name.includes("\0") || name === "..") {
    throw new NatsContextError(`invalid context name: "${name}"`);
  }
  const path = join(baseDir, "context", `${name}.json`);
  let raw: string;
  try {
    raw = await readFile(path, "utf8");
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code === "ENOENT") {
      throw new NatsContextError(`NATS context "${name}" not found at ${path}`);
    }
    throw err;
  }

  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch (err) {
    throw new NatsContextError(
      `NATS context "${name}" is not valid JSON: ${(err as Error).message}`,
    );
  }
  if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed)) {
    throw new NatsContextError(`NATS context "${name}" is not a JSON object`);
  }
  return { name, path, fields: parsed as Record<string, unknown> };
}

/**
 * Resolve a NATS CLI context by name into `NodeConnectionOptions` ready to
 * pass to `connect()`. Pass `"current"` to resolve via `$NATS_CONTEXT` or
 * the `context.txt` selection file.
 */
export async function loadContextOptions(selector: string): Promise<NodeConnectionOptions> {
  return (await loadContextConnectionSnapshot(selector)).connectionOptions;
}

/**
 * Resolve a context once, retaining the selected auth bytes long enough for
 * both the NATS authenticator and an optional sender signer to consume the
 * same snapshot. Not exported from the package root.
 *
 * @internal
 */
export async function loadContextConnectionSnapshot(
  selector: string,
): Promise<ContextConnectionSnapshot> {
  const { name, fields: parsed } = await readContextFile(selector);

  const url = str(parsed["url"]);
  if (!url) throw new NatsContextError(`NATS context "${name}" is missing \`url\``);
  // Parse userinfo just as direct URL mode does. Explicit context auth below
  // wins and removes these fields, so a context can never retain two
  // competing connection credential sources.
  const opts = parseNatsUrl(url);
  let auth: ContextAuthSnapshot =
    opts.user !== undefined || opts.pass !== undefined
      ? { kind: "user-password" }
      : opts.token !== undefined
        ? { kind: "token" }
        : { kind: "none" };

  const creds = str(parsed["creds"]);
  const nkey = str(parsed["nkey"]);
  const userJwt = str(parsed["user_jwt"]);
  const userSeed = str(parsed["user_seed"]);
  if (creds) {
    clearBasicAuth(opts);
    const bytes = await readAuthFile("creds", expandHome(creds));
    const snapshot = new Uint8Array(bytes.buffer, bytes.byteOffset, bytes.byteLength);
    opts.authenticator = credsAuthenticator(snapshot);
    auth = { kind: "creds", bytes: snapshot };
  } else if (nkey) {
    clearBasicAuth(opts);
    // `nats context add` writes `nkey` as a path to a file containing the
    // raw seed. Read it once and pass through `nkeyAuthenticator`, which
    // signs the server nonce on each CONNECT.
    const bytes = await readAuthFile("nkey", expandHome(nkey));
    const snapshot = new Uint8Array(bytes.buffer, bytes.byteOffset, bytes.byteLength);
    opts.authenticator = nkeyAuthenticator(snapshot);
    auth = { kind: "nkey", bytes: snapshot };
  } else if (userJwt) {
    clearBasicAuth(opts);
    // Inline JWT + optional inline seed. When `user_seed` is present
    // alongside `user_jwt` (the typical "decentralised auth without a
    // .creds file" shape), pass the seed bytes so nonce signing works.
    if (userSeed) {
      const snapshot = new TextEncoder().encode(userSeed);
      opts.authenticator = jwtAuthenticator(userJwt, snapshot);
      auth = { kind: "jwt-seed", jwt: userJwt, seed: snapshot };
    } else {
      opts.authenticator = jwtAuthenticator(userJwt);
      auth = { kind: "jwt" };
    }
  } else {
    const token = str(parsed["token"]);
    const user = str(parsed["user"]);
    const password = str(parsed["password"]);
    if (token || user || password) {
      clearBasicAuth(opts);
      if (token) opts.token = token;
      if (user) opts.user = user;
      if (password) opts.pass = password;
      auth = user || password ? { kind: "user-password" } : { kind: "token" };
    }
  }

  try {
    // Optional TLS triple. `nats context add --tlscert/--tlskey/--tlsca`
    // writes file paths; load them into standard node:tls options.
    const cert = str(parsed["cert"]);
    const key = str(parsed["key"]);
    const ca = str(parsed["ca"]);
    const tlsFirst = parsed["tls_first"];
    if (cert || key || ca || tlsFirst === true) {
      const tls: NonNullable<NodeConnectionOptions["tls"]> = {};
      if (cert) tls.cert = await readTlsFile("cert", expandHome(cert));
      if (key) tls.key = await readTlsFile("key", expandHome(key));
      if (ca) tls.ca = await readTlsFile("ca", expandHome(ca));
      if (tlsFirst === true) tls.handshakeFirst = true;
      opts.tls = tls;
    }
  } catch (error) {
    wipeContextAuth(auth);
    throw error;
  }

  const inboxPrefix = str(parsed["inbox_prefix"]);
  if (inboxPrefix) opts.inboxPrefix = inboxPrefix;

  let wiped = false;
  return {
    name,
    connectionOptions: opts,
    auth,
    wipe() {
      if (wiped) return;
      wiped = true;
      wipeContextAuth(auth);
    },
  };
}

function clearBasicAuth(options: NodeConnectionOptions): void {
  delete options.token;
  delete options.user;
  delete options.pass;
}

function wipeContextAuth(auth: ContextAuthSnapshot): void {
  if (auth.kind === "creds" || auth.kind === "nkey") auth.bytes.fill(0);
  if (auth.kind === "jwt-seed") auth.seed.fill(0);
}

/** Expand a leading `~/` to `$HOME/`. */
export function expandHome(path: string): string {
  return path.startsWith("~/") ? join(homedir(), path.slice(2)) : path;
}

async function readTlsFile(field: string, path: string): Promise<string> {
  try {
    return await readFile(path, "utf8");
  } catch (error) {
    throw new NatsContextError(`failed to read TLS ${field} file ${path}`, {
      cause: error,
    });
  }
}

/** Read a credentials / seed file, wrapping filesystem errors in `NatsContextError`. */
export async function readAuthFile(field: string, path: string): Promise<Buffer> {
  try {
    return await readFile(path);
  } catch (error) {
    throw new NatsContextError(`failed to read ${field} file ${path}`, {
      cause: error,
    });
  }
}

/**
 * Parse a NATS URL into `NodeConnectionOptions`, extracting credentials
 * from `userinfo` if present:
 *
 * - `nats://host:port`               → `{ servers: ["nats://host:port"] }`
 * - `nats://TOKEN@host:port`         → `{ servers: [...], token: "TOKEN" }`
 *   (single userinfo component is treated as a token, mirroring the
 *   `nats` CLI's behaviour)
 * - `nats://USER:PASS@host:port`     → `{ servers: [...], user: "USER", pass: "PASS" }`
 * - `tls://...`                      → same shapes, scheme preserved.
 *
 * Comma-separated multi-server URLs (`nats://a:4222,nats://b:4222`) are
 * split and userinfo is only honored if it appears identically on every
 * server — otherwise this function throws (mixed credentials in a single
 * URL is almost certainly a bug; use a NATS context file for that case).
 *
 * Throws if the URL is unparseable, has no host, or uses a non-NATS scheme.
 *
 * @example
 *   import { connect } from "@nats-io/transport-node";
 *   import { parseNatsUrl } from "@synadia-ai/agents";
 *   const nc = await connect(parseNatsUrl("nats://abc123@nats.example.com:4222"));
 */
export function parseNatsUrl(url: string): NodeConnectionOptions {
  const parts = url
    .split(",")
    .map((s) => s.trim())
    .filter((s) => s.length > 0);
  if (parts.length === 0) {
    throw new NatsContextError("empty NATS URL");
  }

  const parsedAll = parts.map((p) => parseSingleNatsUrl(p, url));

  // All servers must agree on userinfo (or all be bare). Mixed userinfo
  // across servers can't be expressed in a single ConnectionOptions.
  const first = parsedAll[0]!;
  for (const p of parsedAll.slice(1)) {
    if (p.token !== first.token || p.user !== first.user || p.pass !== first.pass) {
      throw new NatsContextError(
        `NATS URL has mixed credentials across server entries: ${redactNatsUrl(url)}`,
      );
    }
  }

  const opts: NodeConnectionOptions = {
    servers: parsedAll.map((p) => p.server),
  };
  if (first.token !== undefined) opts.token = first.token;
  if (first.user !== undefined) opts.user = first.user;
  if (first.pass !== undefined) opts.pass = first.pass;
  return opts;
}

interface ParsedNatsUrl {
  server: string; // protocol + host, plus WebSocket path/query (no userinfo)
  token?: string;
  user?: string;
  pass?: string;
}

function parseSingleNatsUrl(part: string, original: string): ParsedNatsUrl {
  // Tolerate scheme-less entries (`host:port`) by prepending nats:// — this
  // mirrors what `@nats-io/transport-node` does internally for `servers`.
  const withScheme = /^[a-z]+:\/\//i.test(part) ? part : `nats://${part}`;

  let parsed: URL;
  try {
    parsed = new URL(withScheme);
  } catch {
    throw new NatsContextError(`invalid NATS URL ${redactNatsUrl(original)}`);
  }
  if (!/^(nats|tls|ws|wss):$/.test(parsed.protocol)) {
    throw new NatsContextError(
      `unsupported scheme "${parsed.protocol}" in NATS URL ${redactNatsUrl(original)}`,
    );
  }
  if (!parsed.host) {
    throw new NatsContextError(`NATS URL ${redactNatsUrl(original)} is missing a host`);
  }

  const websocketSuffix =
    parsed.protocol === "ws:" || parsed.protocol === "wss:"
      ? `${explicitUrlSuffix(withScheme) ? parsed.pathname : ""}${parsed.search}`
      : "";
  const out: ParsedNatsUrl = {
    server: `${parsed.protocol}//${parsed.host}${websocketSuffix}`,
  };

  // WHATWG `URL` squashes both `nats://user@host` and `nats://user:@host`
  // into `password === ""`, losing the distinction between "no separator"
  // (single-component userinfo → token) and "explicit colon, empty
  // password" (user:password form, even if password is empty). Sniff the
  // raw userinfo for a colon to recover the original intent — Python's
  // `urllib.parse.urlparse` distinguishes the two natively (`password is
  // None` vs `password == ""`), but in JS we have to look at the input.
  const userinfoMatch = withScheme.match(/^[a-z]+:\/\/([^/@]*)@/i);
  const hasColonSeparator = (userinfoMatch?.[1] ?? "").includes(":");

  if (hasColonSeparator) {
    // user:password form (decoded; password may be empty if URL was
    // `nats://alice:@host`, but that's still structurally user:password).
    out.user = decodeURIComponent(parsed.username);
    out.pass = decodeURIComponent(parsed.password);
  } else if (parsed.username !== "") {
    // Single userinfo component → token (matches `nats` CLI behaviour).
    out.token = decodeURIComponent(parsed.username);
  }
  return out;
}

/** Whether the raw URL explicitly contained a path or query. */
function explicitUrlSuffix(url: string): boolean {
  const authorityStart = url.indexOf("://") + 3;
  return /[/?]/.test(url.slice(authorityStart));
}

/** Preserve schemes/hosts in diagnostics while replacing every URL userinfo. */
function redactNatsUrl(url: string): string {
  return JSON.stringify(
    url
      .split(",")
      .map((entry) => {
        const trimmed = entry.trim();
        const scheme = trimmed.match(/^[a-z]+:\/\//i)?.[0];
        const authorityStart = scheme?.length ?? 0;
        // Redact through the final `@` even when malformed userinfo contains
        // a literal slash. This may over-redact an `@` in a WebSocket path,
        // which is preferable to leaking credentials from an error path.
        const at = trimmed.lastIndexOf("@");
        return at >= authorityStart
          ? `${trimmed.slice(0, authorityStart)}[redacted]@${trimmed.slice(at + 1)}`
          : trimmed;
      })
      .join(","),
  );
}

function str(v: unknown): string | undefined {
  return typeof v === "string" && v.length > 0 ? v : undefined;
}

function resolveBaseDir(): string {
  const explicit = process.env["NATS_CONFIG_HOME"];
  if (explicit) {
    return explicit.startsWith("~/") ? join(homedir(), explicit.slice(2)) : explicit;
  }
  const xdg = process.env["XDG_CONFIG_HOME"];
  if (xdg) return join(xdg, "nats");
  return join(homedir(), ".config", "nats");
}

async function resolveCurrentName(baseDir: string): Promise<string> {
  const envName = process.env["NATS_CONTEXT"];
  if (envName && envName.length > 0) return envName;
  const path = join(baseDir, "context.txt");
  try {
    const selected = (await readFile(path, "utf8")).trim();
    if (selected.length === 0) {
      throw new NatsContextError(`no NATS context is selected (empty ${path})`);
    }
    return selected;
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code === "ENOENT") {
      throw new NatsContextError(`no NATS context is selected ($NATS_CONTEXT unset, no ${path})`);
    }
    throw err;
  }
}

File: client-sdk/python/src/synadia_ai/agents/context.py
"""``nats`` CLI context loader and URL parser.

Two entry points produce kwargs ready to splat into :func:`nats.connect`:

* :func:`load_context_options` reads context files written by ``nats
  context add`` / ``nats context select`` under ``~/.config/nats`` (or
  ``$NATS_CONFIG_HOME`` / ``$XDG_CONFIG_HOME/nats``).
* :func:`parse_nats_url` parses a single NATS URL and extracts
  credentials from ``userinfo`` if present (token, or user:password).
  ``nats-py``'s ``nats.connect(servers=url)`` does NOT parse userinfo
  on its own — it expects credentials as separate kwargs — but the
  ``nats`` CLI does, which causes a confusing UX gap. Use this helper
  to bridge the two.

Both return a dict you can splat::

    import nats
    from synadia_ai.agents import Agents, load_context_options, parse_nats_url

    nc = await nats.connect(**load_context_options("prod"))
    # or:
    nc = await nats.connect(**parse_nats_url("nats://TOKEN@nats.example.com:4222"))

    agents = Agents(nc=nc)

Mirrors the TS SDK's ``loadContextOptions`` / ``parseNatsUrl`` — same
context-file layout, same auth precedence, same unsupported-field
failures, same URL-parsing semantics. The SDK itself does NOT open
NATS connections; the caller owns ``nc`` and is responsible for
closing it.

Supported context fields: ``url``, ``token``, ``user``/``password``,
``creds`` (with ``~`` expansion), ``user_jwt``, ``inbox_prefix``.

Auth precedence inside a context: ``creds`` > ``user_jwt`` > inline
``token`` / ``user``+``password``.

Unsupported fields raise :class:`~synadia_ai.agents.errors.NatsContextError`
with an actionable message: ``nkey``, TLS triple (``cert`` / ``key`` /
``ca``), ``nsc://...`` URLs.

Ignored (present in JSON but irrelevant to the SDK): ``jetstream_*``,
``socks_proxy``, ``color_scheme``, ``windows_*``, ``user_seed`` (only
meaningful with ``nkey``), ``tls_first``.
"""

from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Any
from urllib.parse import unquote, urlparse

from .errors import NatsContextError

# Fields we know about but do not yet wire into connect kwargs. Presence
# of any of these in a loaded context raises NatsContextError — we do
# NOT silently ignore.
_UNSUPPORTED_FIELDS = ("nkey", "cert", "key", "ca", "nsc")


def load_context_options(selector: str) -> dict[str, Any]:
    """Resolve a ``nats`` CLI context into kwargs for :func:`nats.connect`.

    Pass ``"current"`` to honour ``$NATS_CONTEXT`` → the selection file
    written by ``nats context select``. Any other value loads
    ``<config-dir>/context/<selector>.json`` directly.

    Returns a dict with at minimum ``servers``; auth kwargs (``token``,
    ``user``, ``password``, ``user_credentials``, ``user_jwt_cb``) and
    ``inbox_prefix`` are added when the context declared them.

    Raises :class:`~synadia_ai.agents.errors.NatsContextError` on any failure:
    missing file, malformed JSON, illegal name, missing ``url``,
    unsupported field, missing ``creds`` file, no context selected.
    """
    name = _resolve_current_context_name() if selector == "current" else selector
    _assert_valid_context_name(name)

    path = _context_dir() / f"{name}.json"
    try:
        raw = path.read_text(encoding="utf-8")
    except FileNotFoundError as exc:
        raise NatsContextError(
            f"NATS context {name!r} not found at {path} — try `nats context ls` "
            f"to see which contexts exist, or `nats context add {name} --server=...` "
            "to create one"
        ) from exc

    try:
        parsed = json.loads(raw)
    except json.JSONDecodeError as exc:
        raise NatsContextError(f"NATS context {name!r} is not valid JSON: {exc}") from exc
    if not isinstance(parsed, dict):
        raise NatsContextError(f"NATS context {name!r} must be a JSON object")

    url = parsed.get("url")
    if not isinstance(url, str) or not url:
        raise NatsContextError(f"NATS context {name!r} is missing `url` (or it is empty)")
    servers = _split_urls(url)
    if not servers:
        raise NatsContextError(f"NATS context {name!r}: `url` field resolved to zero servers")

    if any(s.startswith("nsc://") for s in servers):
        raise NatsContextError(
            f"NATS context {name!r}: `nsc://` URLs require shelling out to "
            "`nsc` and are not yet supported"
        )

    for unsupported in _UNSUPPORTED_FIELDS:
        if parsed.get(unsupported):
            raise NatsContextError(
                f"NATS context {name!r}: `{unsupported}` is not yet supported in "
                "synadia-ai-agents; use `creds` / a credentials file if possible, or open "
                "an issue at https://github.com/synadia-ai/synadia-agents/issues"
            )

    out: dict[str, Any] = {"servers": servers}
    out.update(_build_auth_kwargs(name, parsed))

    inbox_prefix = parsed.get("inbox_prefix")
    if isinstance(inbox_prefix, str) and inbox_prefix:
        out["inbox_prefix"] = inbox_prefix

    return out


def _build_auth_kwargs(name: str, parsed: dict[str, Any]) -> dict[str, Any]:
    """Pick the right auth kwargs given the context's auth precedence.

    Precedence: ``creds`` > ``user_jwt`` > inline ``token`` /
    ``user``+``password``. Empty strings are treated as unset so a
    context with one slot zeroed out doesn't shadow another.
    """
    out: dict[str, Any] = {}

    creds = parsed.get("creds")
    if isinstance(creds, str) and creds:
        creds_path = _expand_user(creds)
        if not Path(creds_path).is_file():
            raise NatsContextError(f"NATS context {name!r}: creds file not found: {creds_path}")
        out["user_credentials"] = creds_path
        return out

    user_jwt = parsed.get("user_jwt")
    if isinstance(user_jwt, str) and user_jwt:
        jwt_bytes = user_jwt.encode("utf-8")

        def _jwt_cb() -> bytes:
            return jwt_bytes

        out["user_jwt_cb"] = _jwt_cb
        return out

    for key in ("token", "user", "password"):
        value = parsed.get(key)
        if isinstance(value, str) and value:
            out[key] = value
    return out


# --- internals ----------------------------------------------------------


def _nats_config_dir() -> Path:
    """Resolve the `nats` CLI config dir, honouring the same env priority.

    Priority (matches ``natscli``): ``$NATS_CONFIG_HOME``, then
    ``$XDG_CONFIG_HOME/nats``, then ``$HOME/.config/nats``.
    """
    override = os.environ.get("NATS_CONFIG_HOME")
    if override:
        return Path(_expand_user(override))
    xdg = os.environ.get("XDG_CONFIG_HOME")
    if xdg:
        return Path(xdg) / "nats"
    home = os.environ.get("HOME")
    if home:
        return Path(home) / ".config" / "nats"
    raise NatsContextError(
        "cannot resolve NATS config directory: set $NATS_CONFIG_HOME, $XDG_CONFIG_HOME, or $HOME"
    )


def _context_dir() -> Path:
    return _nats_config_dir() / "context"


def _selection_file() -> Path:
    return _nats_config_dir() / "context.txt"


def _assert_valid_context_name(name: str) -> None:
    """Reject names that would escape the context directory."""
    if not isinstance(name, str) or not name:
        raise NatsContextError("context name must be a non-empty string")
    if "\x00" in name:
        raise NatsContextError(f"context name {name!r} contains a NUL byte")
    if "/" in name or "\\" in name or name == ".." or ".." in name.replace("\\", "/").split("/"):
        raise NatsContextError(f"context name {name!r} contains illegal characters")
    if name.startswith("."):
        raise NatsContextError(f"context name {name!r} must not start with '.'")


def _resolve_current_context_name() -> str:
    """Resolve ``selector="current"`` to a concrete context name.

    ``$NATS_CONTEXT`` wins; otherwise read the selection file written by
    ``nats context select``. Missing both raises :class:`NatsContextError`.
    """
    env_name = os.environ.get("NATS_CONTEXT")
    if env_name:
        return env_name
    selection = _selection_file()
    try:
        selected = selection.read_text(encoding="utf-8").strip()
    except FileNotFoundError as exc:
        raise NatsContextError(
            f"no NATS context is selected ($NATS_CONTEXT unset, no {selection})"
        ) from exc
    if not selected:
        raise NatsContextError(f"no NATS context is selected (empty {selection})")
    return selected


def _split_urls(url: str) -> list[str]:
    """Split a context's ``url`` field into individual server URLs."""
    return [part.strip() for part in url.split(",") if part.strip()]


def _expand_user(path: str) -> str:
    """Expand a leading ``~`` using ``$HOME`` (mirrors TS ``expandTilde``)."""
    if not path.startswith("~"):
        return path
    home = os.environ.get("HOME")
    if not home:
        return path
    if path == "~":
        return home
    if path.startswith("~/") or path.startswith("~\\"):
        return str(Path(home) / path[2:])
    return path


_SUPPORTED_URL_SCHEMES = ("nats", "tls", "ws", "wss")


def parse_nats_url(url: str) -> dict[str, Any]:
    """Parse a NATS URL into ``nats.connect`` kwargs, extracting userinfo.

    Bridges a UX gap between the ``nats`` CLI (which parses userinfo from
    URLs) and ``nats-py`` (which does not). Mirrors the TS SDK's
    :func:`parseNatsUrl`.

    Behaviour:

    * ``nats://host:port`` → ``{"servers": [...]}`` (no auth)
    * ``nats://TOKEN@host:port`` → ``{"servers": [...], "token": ...}``
      — a single userinfo component is treated as a token, mirroring the
      ``nats`` CLI.
    * ``nats://USER:PASS@host:port`` → ``{"servers": [...], "user": ...,
      "password": ...}``
    * ``tls://...``, ``ws://...``, ``wss://...`` schemes preserved on
      output; scheme-less ``host:port`` accepted (treated as
      ``nats://``), matching ``nats-py``'s server-list semantics.
    * Comma-separated multi-server URLs supported when userinfo is
      *identical* across every entry. Mixed userinfo across servers
      cannot be expressed in a single connect-kwargs dict and raises
      :class:`NatsContextError` so the caller fails loudly instead of
      silently dropping all but the first set.
    * URL-decodes userinfo so tokens with reserved chars (``+``, ``@``,
      ``%``) round-trip correctly.

    Raises :class:`NatsContextError` on empty input, unsupported scheme,
    or missing host.

    Example::

        import nats
        from synadia_ai.agents import parse_nats_url

        nc = await nats.connect(**parse_nats_url("nats://abc123@nats.example.com:4222"))
    """
    parts = [p.strip() for p in url.split(",") if p.strip()]
    if not parts:
        raise NatsContextError(f"empty NATS URL: {url!r}")

    parsed_all = [_parse_single_nats_url(p, original=url) for p in parts]

    # All servers must agree on userinfo (or all be bare). Mixed userinfo
    # across servers can't be expressed in one connect-kwargs dict.
    first = parsed_all[0]
    for p in parsed_all[1:]:
        if (
            p.get("token") != first.get("token")
            or p.get("user") != first.get("user")
            or p.get("password") != first.get("password")
        ):
            raise NatsContextError(f"NATS URL has mixed credentials across server entries: {url}")

    out: dict[str, Any] = {"servers": [p["server"] for p in parsed_all]}
    for k in ("token", "user", "password"):
        if k in first:
            out[k] = first[k]
    return out


def _parse_single_nats_url(part: str, *, original: str) -> dict[str, Any]:
    # Tolerate scheme-less entries (`host:port`) by prepending nats://,
    # mirroring nats-py's internal server-list handling.
    with_scheme = part if "://" in part else f"nats://{part}"

    try:
        parsed = urlparse(with_scheme)
    except ValueError as exc:
        raise NatsContextError(f"invalid NATS URL {original!r}: {exc}") from exc

    if parsed.scheme not in _SUPPORTED_URL_SCHEMES:
        raise NatsContextError(f"unsupported scheme {parsed.scheme!r} in NATS URL {original!r}")
    if not parsed.hostname:
        raise NatsContextError(f"NATS URL {original!r} is missing a host")

    # Reconstruct the server URL without userinfo. Re-bracket IPv6 hosts —
    # urlparse strips the brackets but `nats-py` (and most other tools)
    # need them back to disambiguate `host:port`.
    host = parsed.hostname
    host_token = f"[{host}]" if ":" in host else host
    netloc = f"{host_token}:{parsed.port}" if parsed.port is not None else host_token
    server = f"{parsed.scheme}://{netloc}"

    out: dict[str, Any] = {"server": server}
    if parsed.password is not None:
        # user:password (password may be empty if URL is `nats://user:@host`,
        # but we still treat it as the user/password form rather than a token).
        out["user"] = unquote(parsed.username or "")
        out["password"] = unquote(parsed.password)
    elif parsed.username:
        # Single userinfo component → token (matches `nats` CLI behaviour).
        out["token"] = unquote(parsed.username)
    return out


__all__ = ["load_context_options", "parse_nats_url"]
